src/cider/type/plan/FunctionExpr.cpp

```diff
@@ -0,0 +1,307 @@
+ * under the License.
+ */
+
+#include "type/plan/FunctionExpr.h"
+#include "exec/template/IRCodegenUtils.h"
+#include "function/ExtensionFunctionsBinding.h"
+#include "function/ExtensionFunctionsWhitelist.h"
+#include "type/data/sqltypes.h"
+#include "type/plan/UnaryExpr.h"
+
+namespace Analyzer {
+
+void FunctionOper::collect_rte_idx(std::set<int>& rte_idx_set) const {
+  for (unsigned i = 0; i < getArity(); i++) {
+    const auto expr = getArg(i);
+    expr->collect_rte_idx(rte_idx_set);
+  }
+}
+
+void FunctionOper::collect_column_var(
+    std::set<const ColumnVar*, bool (*)(const ColumnVar*, const ColumnVar*)>& colvar_set,
+    bool include_agg) const {
+  for (unsigned i = 0; i < getArity(); i++) {
+    const auto expr = getArg(i);
+    expr->collect_column_var(colvar_set, include_agg);
+  }
+}
+
+std::shared_ptr<Analyzer::Expr> FunctionOper::deep_copy() const {
+  std::vector<std::shared_ptr<Analyzer::Expr>> args_copy;
+  for (size_t i = 0; i < getArity(); ++i) {
+    args_copy.push_back(getArg(i)->deep_copy());
+  }
+  return makeExpr<Analyzer::FunctionOper>(type_info, getName(), args_copy);
+}
+
+bool FunctionOper::operator==(const Expr& rhs) const {
+  if (type_info != rhs.get_type_info()) {
+    return false;
+  }
+  const auto rhs_func_oper = dynamic_cast<const FunctionOper*>(&rhs);
+  if (!rhs_func_oper) {
+    return false;
+  }
+  if (getName() != rhs_func_oper->getName()) {
+    return false;
+  }
+  if (getArity() != rhs_func_oper->getArity()) {
+    return false;
+  }
+  for (size_t i = 0; i < getArity(); ++i) {
+    if (!(*getArg(i) == *(rhs_func_oper->getArg(i)))) {
+      return false;
+    }
+  }
+  return true;
+}
+
+std::string FunctionOper::toString() const {
+  std::string str{"(" + name_ + " "};
+  for (const auto& arg : args_) {
+    str += arg->toString();
+  }
+  str += ")";
+  return str;
+}
+
+std::shared_ptr<Analyzer::Expr> FunctionOperWithCustomTypeHandling::deep_copy() const {
+  std::vector<std::shared_ptr<Analyzer::Expr>> args_copy;
+  for (size_t i = 0; i < getArity(); ++i) {
+    args_copy.push_back(getArg(i)->deep_copy());
+  }
+  return makeExpr<Analyzer::FunctionOperWithCustomTypeHandling>(
+      type_info, getName(), args_copy);
+}
+
+bool FunctionOperWithCustomTypeHandling::operator==(const Expr& rhs) const {
+  if (type_info != rhs.get_type_info()) {
+    return false;
+  }
+  const auto rhs_func_oper =
+      dynamic_cast<const FunctionOperWithCustomTypeHandling*>(&rhs);
+  if (!rhs_func_oper) {
+    return false;
+  }
+  if (getName() != rhs_func_oper->getName()) {
+    return false;
+  }
+  if (getArity() != rhs_func_oper->getArity()) {
+    return false;
+  }
+  for (size_t i = 0; i < getArity(); ++i) {
+    if (!(*getArg(i) == *(rhs_func_oper->getArg(i)))) {
+      return false;
+    }
+  }
+  return true;
+}
+
+JITTypeTag ext_arg_type_to_JIT_type_tag(const ExtArgumentType ext_arg_type) {
+  switch (ext_arg_type) {
+    case ExtArgumentType::Bool:  // pass thru to Int8
```

Comment:
any test cover bool type? ExtArgumentType::Bool equals to JITTypeTag::INT8?

Reply:
We will test other external function in later PR.